src/pair_rann.cpp

```diff
@@ -0,0 +1,1247 @@
+	double dlayerz[jnum][net1.maxlayer];
+	double dlayersumz[jnum][net1.maxlayer];
+	//energy output with forces from analytical derivatives
+	double dsum1;
+	int f = net1.dimensions[0];
+	for (i=0;i<net1.layers-1;i++){
+		for (j=0;j<net1.dimensions[i+1];j++){
+			//energy forward propagation
+			sum[j]=0;
+			for (k=0;k<net1.dimensions[i];k++){
+				if (i==0&&j==0){
+					layer[k]=features[k];
+				}
+				sum[j] += net1.Weights[i][j*net1.dimensions[i]+k]*layer[k];
+			}
+			sum[j] += net1.Biases[i][j];
+			dsum1 = activation[itype][i]->dactivation_function(sum[j]);
+			sum[j] = activation[itype][i]->activation_function(sum[j]);
+			if (i==L-1){
+				energy[j] = sum[j];
+				if (eflag_atom)eatom[i1]=sum[j];
+				if (eflag_global){eng_vdwl +=sum[j];}
+			}
+			//force propagation
+			for (jj=0;jj<jnum;jj++){
+//				if (Bij[jj]==false)continue;
+				dlayersumx[jj][j]=0;
+				dlayersumy[jj][j]=0;
+				dlayersumz[jj][j]=0;
+				for (k=0;k<net1.dimensions[i];k++){
+					if (i==0&&j==0){
+						dlayerx[jj][k]=dfeaturesx[jj*f+k];
+						dlayery[jj][k]=dfeaturesy[jj*f+k];
+						dlayerz[jj][k]=dfeaturesz[jj*f+k];
+					}
+					double w1 = net1.Weights[i][j*net1.dimensions[i]+k];
+					dlayersumx[jj][j] += w1*dlayerx[jj][k];
+					dlayersumy[jj][j] += w1*dlayery[jj][k];
+					dlayersumz[jj][j] += w1*dlayerz[jj][k];
+				}
+				dlayersumx[jj][j] = dsum1*dlayersumx[jj][j];
+				dlayersumy[jj][j] = dsum1*dlayersumy[jj][j];
+				dlayersumz[jj][j] = dsum1*dlayersumz[jj][j];
+				if (i==L-1 && jj < (jnum-1)){
+//					j1 = jlist[jj];
+//					j1 &= NEIGHMASK;
+					j1 = jl[jj];
+					force[j1][0]+=dlayersumx[jj][j];
+					force[j1][1]+=dlayersumy[jj][j];
+					force[j1][2]+=dlayersumz[jj][j];
+				}
+			}
+			if (i==L-1){
+				j1 = ilist[ii];
+				jj = jnum-1;
+				force[j1][0]+=dlayersumx[jj][j];
+				force[j1][1]+=dlayersumy[jj][j];
+				force[j1][2]+=dlayersumz[jj][j];
+			}
+		}
+		//update values for next iteration
+		for (j=0;j<net1.dimensions[i+1];j++){
+			layer[j]=sum[j];
+			for (jj=0;jj<jnum;jj++){
+//				if (Bij[jj]==false)continue;
+				dlayerx[jj][j] = dlayersumx[jj][j];
+				dlayery[jj][j] = dlayersumy[jj][j];
+				dlayerz[jj][j] = dlayersumz[jj][j];
+			}
+		}
+	}
+}
+
+//Called by getproperties. Propagate features and dfeatures through network. Updates force and energy
+void PairRANN::propagateforwardspin(double *features,double *dfeaturesx,double *dfeaturesy,double *dfeaturesz, double *sx, double *sy, double *sz, double * energy,double **force,double **fm,double **virial, int ii,int jnum,int *jl){
+	int i,j,k,jj,j1,itype,i1;
+	int *ilist,*numneigh;
+	ilist = listfull->ilist;
+	int inum = listfull->inum;
+	int *type = atom->type;
+	i1=ilist[ii];
+	itype = map[type[i1]];
+	NNarchitecture net1 = net[itype];
+//	numneigh = listfull->numneigh;
+//	jnum = numneigh[ilist[ii]]+1;//extra value on the end of the array is the self term.
+//	firstneigh = listfull->firstneigh;
+//	jlist = firstneigh[i1];
+	int L = net1.layers-1;
+	double layer[net1.maxlayer];
```

Comment:
these "fortran style" variable length array definitions on the stack are not acceptable and not compliant with C++ standards. They are accepted by GNU C++ compilers only in standard mode with accepts GNU extensions.
Since `jnum` can be large, they must not be allocated on the stack in the first place. Most Linux installations have limits on the stack anyway and thus would cause crashes. Please see the Memory class with its create(), grow(), and destroy() functions. You should also set things up so you avoid frequent re-allocations by storing a `jnummax` variable and then call "memory->grow()" only when jnum is larger than jnummax. there are plenty of example uses of this code construct elsewhere in LAMMPS.

Reply:
I've changed all of these to use the LAMMPS memory class. Unfortunately this seems to slow computation by more than 10% compared to the non-compliant method.